### PR TITLE
module: Voltage Domain driver and related SCMIv3 protocol driver

### PR DESCRIPTION
**Pull request by etienne-lms**
Introduces a basic Voltage Domain driver and a SCMI voltage domain
protocol driver. SCMI voltage domain protocol is introduced in SCMI
specification v3.

Module voltage_domain is a generic HAL for voltage domains, aka power
supply regulators drivers. Based on module configuration, the module
binds each element to the voltage domain backend driver thru a generic
driver interface:
- set & get config, using SCMI voltage domain configuration format.
- set & get voltage level, in microvolt (uV)

Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>

### CODE REVIEW

**Comment by jimqui01** [comment]
jenkins: ok to test

**Comment by jimqui01** [comment]
jenkins: ok to test

**Comment by jimqui01** [comment]
jenkins: ok to test

**Comment by jimqui01** [comment]
Some tab/spaces and documentation issues need to be resolved before merge.

**Comment by etienne-lms** [comment]
Updated with tabs removed and inline description comments added/fixed.

File: module/scmi_voltage_domain/src/mod_scmi_voltage_domain.c [hunk 1]
```diff
@@ -0,0 +1,735 @@
...
+        return status;
+#endif
+
+    status = fwk_module_bind(FWK_ID_MODULE(FWK_MODULE_IDX_VOLTAGE_DOMAIN),
+                             FWK_ID_API(FWK_MODULE_IDX_VOLTAGE_DOMAIN, 0),
+                             &scmi_voltd_ctx.voltd_api);
+    if (status != FWK_SUCCESS)
+        return status;
+
+    for (agent_idx = 0; agent_idx < agent_count; agent_idx++) {
```
Comment (etienne-lms) [inline module/scmi_voltage_domain/src/mod_scmi_voltage_domain.c:698]: the loops below aer usless. Will remove.

File: module/scmi_voltage_domain/src/mod_scmi_voltage_domain.c [hunk 2]
```diff
@@ -0,0 +1,735 @@
...
+    const struct mod_voltd_api *voltd_api;
+
+    /* Number of domain devices */
+    int voltd_devices;
+
+    /* Pointer to a table of domain operations */
+    struct voltd_operations *voltd_ops;
+
+#ifdef BUILD_HAS_RESOURCE_PERMISSIONS
+    /* SCMI Resource Permissions API (TODO) */
```
Comment (etienne-lms) [inline module/scmi_voltage_domain/src/mod_scmi_voltage_domain.c:66]: remove `(TODO)`.

File: module/scmi_voltage_domain/src/mod_scmi_voltage_domain.c [hunk 3]
```diff
@@ -0,0 +1,735 @@
...
+    uint32_t config = 0;
+    int status = 0;
+
+    inmsg = (const struct scmi_voltd_attributes_a2p *)(void *)payload;
+
+    status = get_device(service_id, inmsg->domain_id, &device, NULL);
+    if (status != FWK_SUCCESS)
+        goto exit;
+
+    // FIXME: who knows the name? scmi_voltd, voltd, platform regu driver?
```
Comment (etienne-lms) [inline module/scmi_voltage_domain/src/mod_scmi_voltage_domain.c:310]: remove inline comment

**Comment by etienne-lms** [comment]
Updated:
- Remove useless loop in scmi_voltd_bind()
- Fix/remove inline comments.
- Simplify scmi_voltd_permissions_handler().

**Comment by etienne-lms** [comment]
Updated: fixed remaining tabulation and inline comments.

**Comment by jimqui01** [comment]
jenkins: ok to test

**Comment by jimqui01** [comment]
jenkins: ok to test

**Review by jimqui01 — APPROVED** [review]

**Comment by etienne-lms** [comment]
thanks

**Pull request merged**